File: backend/app/routers/life.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.db import get_db
from app.deps import get_current_user
from app.models import LifeRecord, User
from app.schemas import LifeRecordCreateIn, LifeRecordUpdateIn

router = APIRouter(prefix='/life-records', tags=['life-records'])


def success(data=None, message='success'):
  return {'code': 0, 'message': message, 'data': data}


def serialize_record(item: LifeRecord):
  return {
    'id': item.id,
    'recordType': item.record_type,
    'serviceName': item.service_name,
    'title': item.title,
    'amount': item.amount,
    'eventDate': item.event_date,
    'remark': item.remark,
    'status': item.status,
    'createdAt': item.created_at.strftime('%Y-%m-%d %H:%M')
  }


@router.get('')
def get_life_records(
  record_type: str = Query(default='ledger'),
  keyword: str = Query(default=''),
  db: Session = Depends(get_db),
  user: User = Depends(get_current_user)
):
  query = db.query(LifeRecord).filter(
    LifeRecord.user_id == user.id,
    LifeRecord.record_type == record_type
  )
  if keyword:
    query = query.filter(LifeRecord.title.contains(keyword))
  items = query.order_by(LifeRecord.id.desc()).all()
  total_amount = sum(item.amount or 0 for item in items)
  return success({
    'list': [serialize_record(item) for item in items],
    'summary': {
      'count': len(items),
      'totalAmount': round(total_amount, 2)
    }
  })


@router.post('')
def create_life_record(
  payload: LifeRecordCreateIn,
  db: Session = Depends(get_db),
  user: User = Depends(get_current_user)
):
  item = LifeRecord(
    user_id=user.id,
    record_type=payload.recordType,
    service_name=payload.serviceName,
    title=payload.title,
    amount=payload.amount or 0,
    event_date=payload.eventDate,
    remark=payload.remark,
    status=payload.status or 'submitted'
  )
  db.add(item)
  db.commit()
  db.refresh(item)
  return success(serialize_record(item))


@router.put('/{record_id}')
def update_life_record(
  record_id: int,
  payload: LifeRecordUpdateIn,
  db: Session = Depends(get_db),
  user: User = Depends(get_current_user)
):
  item = db.query(LifeRecord).filter(
    LifeRecord.id == record_id,
    LifeRecord.user_id == user.id
  ).first()
  if not item:
    return success(None, '记录不存在')

  if payload.serviceName is not None:
    item.service_name = payload.serviceName
  if payload.title is not None:
    item.title = payload.title
  if payload.amount is not None:
    item.amount = payload.amount
  if payload.eventDate is not None:
    item.event_date = payload.eventDate
  if payload.remark is not None:
    item.remark = payload.remark
  if payload.status is not None:
    item.status = payload.status

  db.add(item)
  db.commit()
  db.refresh(item)
  return success(serialize_record(item))


@router.delete('/{record_id}')
def delete_life_record(
  record_id: int,
  db: Session = Depends(get_db),
  user: User = Depends(get_current_user)
):
  item = db.query(LifeRecord).filter(
    LifeRecord.id == record_id,
    LifeRecord.user_id == user.id
  ).first()
  if not item:
    return success(False, '记录不存在')
  db.delete(item)
  db.commit()
  return success(True)

File: frontend/api/life.js
import { USE_MOCK_API, STORAGE_KEYS } from '../utils/config'
import { getStorage, setStorage } from '../utils/storage'
import { mockLifeRecords } from '../mock/data'
import { del, get, post, put } from './request'
import { formatDateTimeLabel } from '../utils/format'

function getLocalRecords() {
  const list = getStorage(STORAGE_KEYS.lifeRecords, [])
  if (list && list.length) {
    return list
  }
  setStorage(STORAGE_KEYS.lifeRecords, mockLifeRecords)
  return mockLifeRecords
}

function saveLocalRecords(list) {
  setStorage(STORAGE_KEYS.lifeRecords, list)
}

function buildSummary(list) {
  const totalAmount = list.reduce((sum, item) => sum + Number(item.amount || 0), 0)
  return {
    count: list.length,
    totalAmount: Number(totalAmount.toFixed(2))
  }
}

export async function getLifeRecords(params = {}) {
  const recordType = params.recordType || 'ledger'
  const keyword = params.keyword || ''

  if (USE_MOCK_API) {
    let list = getLocalRecords().filter((item) => item.recordType === recordType)
    if (keyword) {
      list = list.filter((item) => item.title.indexOf(keyword) > -1)
    }
    return {
      list,
      summary: buildSummary(list)
    }
  }

  try {
    const response = await get('/life-records', {
      record_type: recordType,
      keyword
    })
    return response.data
  } catch (error) {
    let list = getLocalRecords().filter((item) => item.recordType === recordType)
    if (keyword) {
      list = list.filter((item) => item.title.indexOf(keyword) > -1)
    }
    return {
      list,
      summary: buildSummary(list)
    }
  }
}

export async function createLifeRecord(payload) {
  if (USE_MOCK_API) {
    const list = getLocalRecords()
    const item = {
      id: Date.now(),
      createdAt: formatDateTimeLabel(new Date()),
      status: payload.status || 'submitted',
      ...payload
    }
    saveLocalRecords([item].concat(list))
    return item
  }

  try {
    const response = await post('/life-records', payload)
    return response.data
  } catch (error) {
    const list = getLocalRecords()
    const item = {
      id: Date.now(),
      createdAt: formatDateTimeLabel(new Date()),
      status: payload.status || 'submitted',
      ...payload
    }
    saveLocalRecords([item].concat(list))
    return item
  }
}

export async function updateLifeRecord(id, payload) {
  if (USE_MOCK_API) {
    const list = getLocalRecords().map((item) => {
      if (item.id === id) {
        return {
          ...item,
          ...payload
        }
      }
      return item
    })
    saveLocalRecords(list)
    return true
  }

  try {
    const response = await put('/life-records/' + id, payload)
    return response.data
  } catch (error) {
    const list = getLocalRecords().map((item) => {
      if (item.id === id) {
        return {
          ...item,
          ...payload
        }
      }
      return item
    })
    saveLocalRecords(list)
    return true
  }
}

export async function deleteLifeRecord(id) {
  if (USE_MOCK_API) {
    const list = getLocalRecords().filter((item) => item.id !== id)
    saveLocalRecords(list)
    return true
  }

  try {
    const response = await del('/life-records/' + id)
    return response.data
  } catch (error) {
    const list = getLocalRecords().filter((item) => item.id !== id)
    saveLocalRecords(list)
    return true
  }
}
